zkevm-circuits/src/evm_circuit/util.rs

```diff
@@ -62,6 +69,232 @@ impl<F: FieldExt> Expr<F> for &Cell<F> {
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct CellManager<F> {
+    width: usize,
+    height: usize,
+    cells: Vec<Cell<F>>,
+    columns: Vec<CellColumn<F>>,
+}
+
+impl<F: FieldExt> CellManager<F> {
+    pub(crate) fn new(
+        meta: &mut ConstraintSystem<F>,
+        width: usize,
+        height: usize,
+        advices: &[Column<Advice>],
+        height_offset: usize,
+    ) -> Self {
+        // Setup the columns and query the cells
+        assert_eq!(advices.len(), width);
+        let mut cells = Vec::with_capacity(height * width);
+        let mut columns = Vec::with_capacity(height);
+        meta.create_gate("Query rows for step", |meta| {
+            for c in 0..width {
+                for r in 0..height {
+                    cells.push(Cell::new(meta, advices[c], height_offset + r));
+                }
+                columns.push(CellColumn {
+                    index: c,
+                    cell_type: CellType::Storage,
+                    height: 0,
+                    expr: cells[c * height].expr(),
+                });
+            }
+            vec![0.expr()]
+        });
+
+        // Mark columns used for lookups
+        let mut column_idx = 0;
+        for &(table, count) in LOOKUP_CONFIG {
+            for _ in 0usize..count {
+                columns[column_idx].cell_type = CellType::Lookup(table);
+                column_idx += 1;
+            }
+        }
+
+        Self {
+            width,
+            height,
+            cells,
+            columns,
+        }
+    }
+
+    pub(crate) fn query_cells(&mut self, cell_type: CellType, count: usize) -> Vec<Cell<F>> {
+        let mut cells = Vec::with_capacity(count);
+        while cells.len() < count {
+            let column_idx = self.next_column(cell_type);
+            let column = &mut self.columns[column_idx];
+            cells.push(self.cells[column_idx * self.height + column.height].clone());
+            column.height += 1;
+        }
+        cells
+    }
+
+    pub(crate) fn query_cell(&mut self, cell_type: CellType) -> Cell<F> {
+        self.query_cells(cell_type, 1)[0].clone()
+    }
+
+    fn next_column(&self, cell_type: CellType) -> usize {
+        let mut best_index: Option<usize> = None;
+        let mut best_height = self.height;
+        for column in self.columns.iter() {
+            if column.cell_type == cell_type && column.height < best_height {
+                best_index = Some(column.index);
+                best_height = column.height;
+            }
+        }
+        match best_index {
+            Some(index) => index,
+            None => unreachable!(format!("not enough cells for query: {:?}", cell_type)),
+        }
+    }
+
+    pub(crate) fn get_height(&self) -> usize {
+        self.columns
+            .iter()
+            .map(|column| column.height)
+            .max()
+            .unwrap()
+    }
+
+    pub(crate) fn get_stats(&self) -> BTreeMap<CellType, (usize, usize, usize)> {
```

Comment:
So the main reason for this function to exist is to get some insights in what needs to be stored. Perhaps in the future something smarter can be done with this (like automatically find the optimal number of columns for each cell type), but for now it can be used to just print out the info to see what's going on and what is used. I'm using a `BTreeMap` here because otherwise if you run over the keys/values the order is always different and so it's harder to compare the results of different calls to this function without having to sort on the keys again. Hope this makes sense!

Reply:
Makes sense. 

If the number of accesses is not big it's fine to use the `BTreeMap`.
If we call this a lot of times, then we need to take into account that `BTreeMap` storage is slower. So depending on the frequency which with this code is called, maybe we can try to think about better ways in the future :smile: